freshmaker/handlers/koji/rebuild_images_on_rpm_advisory_change.py

```diff
@@ -303,24 +303,33 @@ def _record_batches(self, batches, event, builds=None):
                     # in a way that we have to pass all the ODCS composes to it or
                     # no ODCS compose at all.
                     if image["generate_pulp_repos"] or not image["published"]:
-                        # Add arch specific composes
-                        for arch_content_sets in image["multi_arch_content_sets"].values():
-                            cache_key = " ".join(sorted(arch_content_sets))
-                            if cache_key in image["compose_sources"]:
-                                # This compose has same 'source' value as one of existing
-                                # composes in original build, freshmaker doesn't need to
-                                # generate this compose, since that one will be renewed by
-                                # osbs automatically when it's included in 'compose_ids'
-                                # in build task params
+                        original_pulp_compose_sources = set()
+                        for compose_id in image["original_odcs_compose_ids"]:
+                            compose = self.odcs.get_compose(compose_id)
+                            source_type = compose.get("source_type")
+                            # source_type of pulp composes is 4
+                            if source_type != 4:
                                 continue
-
-                            # Check if the compose for these content_sets is
+                            source_value = compose.get("source", "")
+                            for source in source_value.split():
+                                original_pulp_compose_sources.add(source.strip())
+
+                        # Add content set to new_pulp_sources if it's not found
+                        # in original_pulp_compose_sources
+                        new_pulp_sources = set()
+                        for content_set in image["content_sets"]:
+                            if content_set not in original_pulp_compose_sources:
+                                new_pulp_sources.add(content_set)
+
+                        if new_pulp_sources:
+                            # Check if the compose for these new pulp sources is
                             # already cached and use it in this case.
+                            cache_key = " ".join(sorted(new_pulp_sources))
                             if cache_key in odcs_cache:
                                 db_compose = odcs_cache[cache_key]
                             else:
                                 compose = self.odcs.prepare_pulp_repo(
-                                    build, arch_content_sets)
+                                    build, list(new_pulp_sources))
```

Comment:
I don't understand composes completely, but if we are creating new compose with only NEW pulp sources, so the old ones will stay or they will be replaced by new ones or how this works?

Reply:
We only generate new pulp composes for the contest sets which don't exist in original odcs composes, the original odcs composes will be renewed, so in the image build task, the composes are old ones and the new ones added here, the purpose of this PR is to remove the duplicated repos in old composes, we already has 4694dda to not adding duplicated repos, but we can't remove the existing duplicated repos without this PR. Actually this is a temporary solution and probably can be removed after a while when there is no latest published images are having duplicated repos enabled.

Check:
https://github.com/redhat-exd-rebuilds/freshmaker/blob/040cd4b8f4feb4d7607e47d20bd106ef6c7c7e01/freshmaker/handlers/koji/rebuild_images_on_rpm_advisory_change.py#L293
Note: the `odcs_compose_ids` has been changed to `original_odcs_compose_ids` in this PR.

Reply:
Thx for explanation